lib/spack/spack/url_buildcache.py

```diff
@@ -0,0 +1,1262 @@
+            record.checksum_alg,
+            record.checksum[:2],
+            record.checksum,
+        )
+
+        web_util.push_to_url(blob_path, blob_destination_url, keep_original=False)
+
+    @classmethod
+    def push_manifest(
+        cls,
+        mirror_url: str,
+        manifest_name: str,
+        manifest: BuildcacheManifest,
+        tmpdir: str,
+        component_type: BuildcacheComponent = BuildcacheComponent.SPEC,
+        signing_key: Optional[str] = None,
+    ) -> None:
+        """Given a BuildcacheManifest, push it to the mirror using the given manifest
+        name.  The component_type is used to indicate what type of thing the manifest
+        represents, so it can be placed in the correct relative path within the mirror.
+        If a signing_key is provided, it will be used to clearsign the manifest before
+        pushing it."""
+        # write the manifest to a temporary location
+        manifest_file_name = f"{manifest_name}.manifest.json"
+        manifest_path = os.path.join(tmpdir, manifest_file_name)
+        with open(manifest_path, "w", encoding="utf-8") as f:
+            json.dump(manifest.to_dict(), f, indent=0, separators=(",", ":"))
+            # Note: when using gpg clear sign, we need to avoid long lines (19995
+            # chars). If lines are longer, they are truncated without error. So,
+            # here we still add newlines, but no indent, so save on file size and
+            # line length.
+
+        if signing_key:
+            manifest_path = sign_file(signing_key, manifest_path)
+
+        manifest_destination_url = url_util.join(
+            mirror_url, *cls.get_relative_path_components(component_type), manifest_file_name
+        )
+
+        web_util.push_to_url(manifest_path, manifest_destination_url, keep_original=False)
+
+    @classmethod
+    def push_local_file_as_blob(
+        cls,
+        local_file_path: str,
+        mirror_url: str,
+        manifest_name: str,
+        component_type: BuildcacheComponent,
+        compression: str = "none",
+    ) -> None:
+        """Convenience method to push a local file to a mirror as a blob.  Both manifest
+        and blob are pushed as a component of the given component_type.  If compression
+        is 'gzip' the blob will be compressed before pushing, otherwise it will be pushed
+        uncompressed."""
+        cache_class = get_url_buildcache_class()
+        checksum_algo = "sha256"
+        blob_to_push = local_file_path
+        tmpdir = tempfile.mkdtemp()
+
+        try:
+            blob_to_push = os.path.join(tmpdir, os.path.basename(local_file_path))
+
+            with compression_writer(blob_to_push, compression, checksum_algo) as (
+                compressor,
+                checker,
+            ):
+                with open(local_file_path, "r", encoding="utf-8") as fd:
+                    compressor.write(fd.read().encode("utf-8"))
```

Comment:
Copy the file verbatim, decoding it is redundant/limiting/error prone; don't read it entirely into memory.

```suggestion
            with compression_writer(blob_to_push, compression, checksum_algo) as (
                fout,
                checker,
            ), open(local_file_path, "rb") as fin:
                shutil.copyfileobj(fin, fout)
```